gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,382 @@
+	 * @param result set populated with objects near from the frustum
+	 */
+	public ObjectSet<T> query (Frustum frustum, ObjectSet<T> result) {
+		root.query(frustum, result);
+		return result;
+	}
+
+	public T rayCast (Ray ray, RayCastResult<T> result) {
+		result.distance = result.maxDistanceSq;
+		root.rayCast(ray, result);
+		return result.geometry;
+	}
+
+	/** Method to get nodes as bounding boxes. Useful for debug purpose.
+	 *
+	 * @param boxes
+	 */
+	public ObjectSet<BoundingBox> getNodesBoxes (ObjectSet<BoundingBox> boxes) {
+		root.getBoundingBox(boxes);
+		return boxes;
+	}
+
+	protected class OctreeNode {
+
+		private int level;
+
+		// Represents the boundary of the node
+		private final BoundingBox bounds;
+		private final Array<OctreeNode> children = new Array<>(8);
+		private final Array<T> geometries = new Array<T>(maxItemsPerNode);
+
+		public OctreeNode () {
+			bounds = new BoundingBox();
+		}
+
+		private void split () {
+			float midx = (bounds.max.x + bounds.min.x) * 0.5f;
+			float midy = (bounds.max.y + bounds.min.y) * 0.5f;
+			float midz = (bounds.max.z + bounds.min.z) * 0.5f;
+
+			int deeperLevel = level - 1;
+
+			children.add(createNode(new Vector3(bounds.min.x, midy, midz), new Vector3(midx, bounds.max.y, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, midy, midz), new Vector3(bounds.max.x, bounds.max.y, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, midy, bounds.min.z), new Vector3(bounds.max.x, bounds.max.y, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, midy, bounds.min.z), new Vector3(midx, bounds.max.y, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, bounds.min.y, midz), new Vector3(midx, midy, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, bounds.min.y, midz), new Vector3(bounds.max.x, midy, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, bounds.min.y, bounds.min.z), new Vector3(bounds.max.x, midy, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, bounds.min.y, bounds.min.z), new Vector3(midx, midy, midz), deeperLevel));
+
+			// Move geometries from parent to children
+			for (OctreeNode child : children) {
+				for (T geometry : this.geometries) {
+					child.add(geometry);
+				}
+			}
+			this.geometries.clear();
+		}
+
+		private void merge() {
+			for (OctreeNode node : children) {
+				geometries.addAll(node.geometries);
+				freeNode(node);
+			}
+			children.clear();
+		}
+
+		private OctreeNode createNode(Vector3 min, Vector3 max, int level) {
+			OctreeNode node = nodePool.obtain();
+			node.bounds.set(min, max);
+			node.level = level;
+			return node;
+		}   
+
+		private void freeNode(OctreeNode node) {
+			node.geometries.clear();
+			node.children.clear();
+			nodePool.free(node);
+		}
+
+		protected boolean contains (Collider<T> boundingBoxHandler, T geometry) {
```

Comment:
not sure to understand what this method is and what it's used for.. ?
also, contains / intersects is a bit confusing.

Reply:
Nice catch, this method is not necessary anymore.